assets/js/modules/ads/components/setup/SetupMainPAX.js

```diff
@@ -0,0 +1,234 @@
+/**
+ * Ads Main PAX setup component.
+ *
+ * Site Kit by Google, Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	createInterpolateElement,
+	Fragment,
+	useCallback,
+} from '@wordpress/element';
+import { __, _x } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import AdsIcon from '../../../../../svg/graphics/ads.svg';
+import SetupForm from './SetupForm';
+import SupportLink from '../../../../components/SupportLink';
+import AdBlockerWarning from '../common/AdBlockerWarning';
+import { SpinnerButton } from 'googlesitekit-components';
+import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
+import { CORE_LOCATION } from '../../../../googlesitekit/datastore/location/constants';
+import { ADWORDS_SCOPE, MODULES_ADS } from '../../datastore/constants';
+import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
+import useQueryArg from '../../../../hooks/useQueryArg';
+import { addQueryArgs } from '@wordpress/url';
+import PAXEmbeddedApp from '../PAXEmbeddedApp';
+const { useSelect, useDispatch, useRegistry } = Data;
+
+export default function SetupMainPAX( { finishSetup } ) {
+	const [ showPaxApp, setShowPaxApp ] = useQueryArg( 'pax' );
+
+	const registry = useRegistry();
+
+	const isAdBlockerActive = useSelect( ( select ) =>
+		select( CORE_USER ).isAdBlockerActive()
+	);
+	const hasAdwordsScope = useSelect( ( select ) =>
+		select( CORE_USER ).hasScope( ADWORDS_SCOPE )
+	);
+
+	const redirectURL = addQueryArgs( global.location.href, {
+		pax: 1,
+		notification: 'ads', // Arbitrary value to prevent authentication_success from surfacing.
+	} );
+
+	const oAuthURL = useSelect( ( select ) =>
+		select( CORE_USER ).getConnectURL( {
+			additionalScopes: [ ADWORDS_SCOPE ],
+			redirectURL,
+		} )
+	);
+	const isNavigatingToOAuthURL = useSelect( ( select ) => {
+		if ( ! oAuthURL ) {
+			return false;
+		}
+
+		return select( CORE_LOCATION ).isNavigatingTo( oAuthURL );
+	} );
+
+	const { navigateTo } = useDispatch( CORE_LOCATION );
+	const { setPaxConversionID, setExtCustomerID, submitChanges } =
+		useDispatch( MODULES_ADS );
+
+	const onLaunch = useCallback(
+		async ( app ) => {
+			if ( app && typeof app?.getServices === 'function' ) {
```

Comment:
This seems like it's doing a bit more than necessary. Why do all of this in the `onLaunch` handler? We only need to store the app instance, or maybe just the services.

Everything else here should be bound to the handler for the main CTA to complete setup.

Reply:
This is done as part of another issue, that is waiting on this one to be merged, #6560, it is setting the app instance only, and rest of logic is moved as part of new function. In this issue, IB is referencing all this operation in `onLaunch` callback.

Reply:
That issue (#8560) is related of course but my comment here is still relevant. 

> In this issue, IB is referencing all this operation in `onLaunch` callback.

Please review the IB again :)

The relevant points from the IB are:
> * When rendering `PAXEmbeddedApp` also pass an `onLaunch` function prop to receive a reference to the `app` instance on successful launch (for use later)

Storing this reference is the only thing that should happen in this function. The reference is then used later, on demand.

> Add the `SpinnerButton` from the main setup form below the PAX app
> * **When clicked**, use the `app` instance returned the PAX `onLaunch` to get access to the Ads services (via `app.getServices()` ([example](https://github.com/google/partner-ads-experience/tree/main?tab=readme-ov-file&rgh-link-date=2024-04-15T21%3A55%3A52Z#get-services-supported-by-the-ads-app)))

This button is present below the create account CTA, but should still be there when PAX is launched, and is the trigger to set settings and submit changes as described in the IB. We should not be calling submit changes, etc in the `onLaunch` – this is why you're having problems with `finishSetup` being called too early.